st3/lsp_utils/_client_handler/abstract_plugin.py

```diff
@@ -0,0 +1,161 @@
+
+    def on_notification(self, method: str, handler: Callable[[Any], None]) -> None:
+        setattr(self.__plugin, method2attr(method), lambda params: handler(params))
+
+    def on_request(self, method: str, handler: Callable[[Any, Callable[[Any], None]], None]) -> None:
+        def send_response(request_id, result):
+            session = self.__plugin.weaksession()
+            if session:
+                session.send_response(Response(request_id, result))
+
+        def on_response(params, request_id):
+            handler(params, lambda result: send_response(request_id, result))
+
+        setattr(self.__plugin, method2attr(method), on_response)
+
+    def send_notification(self, method: str, params: Any) -> None:
+        session = self.__plugin.weaksession()
+        if session:
+            session.send_notification(Notification(method, params))
+
+    def send_request(self, method: str, params: Any, handler: Callable[[Any, bool], None]) -> None:
+        session = self.__plugin.weaksession()
+        if session:
+            session.send_request(
+                Request(method, params), lambda result: handler(result, False), lambda result: handler(result, True))
+        else:
+            handler(None, True)
+
+
+class ClientHandler(AbstractPlugin, ClientHandlerInterface):
+    """
+    The base class for creating an LSP plugin.
+    """
+
+    # --- AbstractPlugin handlers -------------------------------------------------------------------------------------
+
+    @classmethod
+    def name(cls) -> str:
+        return cls.get_displayed_name()
+
+    @classmethod
+    def configuration(cls) -> Tuple[sublime.Settings, str]:
+        return cls.read_settings()
+
+    @classmethod
+    def additional_variables(cls) -> Optional[Dict[str, str]]:
+        return cls.get_additional_variables()
+
+    @classmethod
+    def needs_update_or_installation(cls) -> bool:
+        if cls.manages_server():
+            server = cls.get_server()
+            return bool(server) and server.needs_installation()
+        return False
+
+    @classmethod
+    def install_or_update(cls) -> None:
+        server = cls.get_server()
+        if server:
+            server.install_or_update()
+
+    @classmethod
+    def can_start(cls, window: sublime.Window, initiating_view: sublime.View,
+                  workspace_folders: List[WorkspaceFolder], configuration: ClientConfig) -> Optional[str]:
+        if cls.manages_server():
+            server = cls.get_server()
+            if not server or server.get_status() == ServerStatus.ERROR:
+                return "{}: Error installing server dependencies.".format(cls.package_name)
+            if server.get_status() != ServerStatus.READY:
+                return "{}: Server installation in progress...".format(cls.package_name)
+        message = cls.is_allowed_to_start(window, initiating_view, workspace_folders, configuration)
```

Comment:
While trying to fix https://github.com/sublimelsp/LSP/issues/1440, I learned that if you listen for a change in a sublime.Settings object, then your plugin should never use `.set()` on that object as it can get into an infinite recursion.

Currently I'm guilty of doing that in a bunch of my helper packages. I will have to change that. I think the best place for these kind of changes is actually here in `can_start` (or `is_allowed_to_start` in this interface), and to do the mutation on the ClientConfig object.

Unfortunately LSP-pyright now also mutates its sublime.Settings object to dynamically change the extraPaths. What's also unfortunate is that the old `on_start` doesn't receive enough information.

Reply:
> Unfortunately LSP-pyright now also mutates its sublime.Settings object to dynamically change the extraPaths. 

what if revise configuration in `on_client_configuration_ready`? I do it there originally.

Reply:
> While trying to fix [sublimelsp/LSP#1440](https://github.com/sublimelsp/LSP/issues/1440), I learned that if you listen for a change in a sublime.Settings object, then your plugin should never use `.set()` on that object as it can get into an infinite recursion.

That doesn't sound right. I guess the more precise way to say it would be "you can get into an infinite loop if you modify your settings in response to the change listener. I don't currently understand in what circumstances that would happen.

Reply:
> > Unfortunately LSP-pyright now also mutates its sublime.Settings object to dynamically change the extraPaths.
> 
> what if revise configuration in `on_client_configuration_ready`? I do it there originally.

Before we ask that question, maybe we should first ask what is the expected flow on the user changing some plugin setting.

Current settings reading is done once at plugin init so whether we use `on_client_configuration_ready` or `on_settings_read` it will not work correctly as those won't re-run on changing settings.

Reply:
Well, they can be re-run provided that LSP calls `plugin.configuration()` on each setting change, but that causes the infinite recursion for some plugins.

So we really shouldn't be using sublime.Settings.set anywhere. So the only place (I think?) to modify things is in ClientConfig, which is passed in `can_start`.

Reply:
So something like this for pyright could work:

```python
    @classmethod
    def is_allowed_to_start(cls, window, initiating_view, workspace_folders, configuration):
        if configuration and settings.get("dev_environment") == "sublime_text":
            paths = configuration.settings.get("python.analysis.extraPaths")
            if not isinstance(paths, list):
                paths = []
            paths.extend(cls.find_package_dependency_dirs())
            configuration.settings.set("python.analysis.extraPaths", paths)
```

this way we don't change a sublime.Settings object but we still modify the server settings (just in time).

Reply:
but it would only work for the ST4 branch...

Reply:
although you can add a version check in `on_settings_read`: `if sublime.version() < "4000": ...`, so that would make it work for both ST3 and ST4 :)

Reply:
> So we really shouldn't be using sublime.Settings.set anywhere. So the only place (I think?) to modify things is in ClientConfig, which is passed in `can_start`.

There was a use case where we actually wanted to change the settings and even save them to disk. It was for migrating old language settings to the new one.

> So something like this for pyright could work:

This class method will be called multiple times though (once per window at least). Would the plugin have to make sure that it doesn't add redundant items (when extending array) or "configuration" be pristine on each call?

BTW. What about keeping current semantics where the client is allowed to modify and save settings and adding some extra API to be called when settings change?

BTW2. As I've said before, the semantics of `can_start` doesn't really make it clear that you are allowed to modify the client config from it. If we are changing API, maybe there should be a new method added specifically for the purpose of allowing the client to mutate the config.

Reply:
> BTW. What about keeping current semantics where the client is allowed to modify and save settings and adding some extra API to be called when settings change?

Agreed.

> BTW2. As I've said before, the semantics of can_start doesn't really make it clear that you are allowed to modify the client config from it. If we are changing API, maybe there should be a new method added specifically for the purpose of allowing the client to mutate the config.

Agreed. I'll try to work on this.

Reply:
I don't think we have arrived at the correct solution yet in LSP.

Either we merge this refactor as is or wait until we figure out something better for LSP.

But it's really a separate issue, I would say.